src/dft/backends/mklcpu/commit.cpp

```diff
@@ -31,113 +31,161 @@
+    DFT_ERROR status[2] = { DFTI_BAD_DESCRIPTOR };
+
+    for (auto dir : { DIR::fwd, DIR::bwd }) {
+        if (config_values.dimensions.size() == 1)
+            status[dir] = DftiCreateDescriptor(&bidirection_handle[dir], mklcpu_prec, mklcpu_dom, 1,
+                                               config_values.dimensions[0]);
+        else
+            status[dir] = DftiCreateDescriptor(&bidirection_handle[dir], mklcpu_prec, mklcpu_dom,
+                                               config_values.dimensions.size(),
+                                               config_values.dimensions.data());
     }
 
-    void commit(const detail::dft_values<prec, dom>& config_values) override {
-        set_value(handle, config_values);
-        auto status = DftiCommitDescriptor(handle);
-        if (status != DFTI_NO_ERROR) {
-            throw oneapi::mkl::exception(
-                "dft/backends/mklcpu", "commit",
-                "DftiCommitDescriptor failed with status: " + std::to_string(status));
-        }
+    if (status[0] != DFTI_NO_ERROR && status[1] != DFTI_NO_ERROR) {
+        std::string err = std::string("DftiCreateDescriptor failed with status : ") +
+                          DftiErrorMessage(status[0]) + std::string(", ") +
+                          DftiErrorMessage(status[1]);
+        throw oneapi::mkl::exception("dft/backends/mklcpu", "create_descriptor", err);
     }
+}
 
-    virtual void* get_handle() noexcept override {
-        return handle;
-    }
+template <dft::detail::precision prec, dft::detail::domain dom>
+commit_derived_impl<prec, dom>::~commit_derived_impl() {
+    for (auto dir : { DIR::fwd, DIR::bwd })
+        DftiFreeDescriptor(&bidirection_handle[dir]);
+}
 
-    virtual ~commit_derived_impl() override {
-        DftiFreeDescriptor((DFTI_DESCRIPTOR_HANDLE*)&handle);
-    }
+template <dft::detail::precision prec, dft::detail::domain dom>
+void commit_derived_impl<prec, dom>::commit(
+    const dft::detail::dft_values<prec, dom>& config_values) {
+    set_value(bidirection_handle.data(), config_values);
+
+    this->get_queue()
+        .submit([&](sycl::handler& cgh) {
+            auto bidir_handle_obj =
+                bidirection_buffer.template get_access<sycl::access::mode::read_write>(cgh);
+
+            host_task<detail::kernel_name<mklcpu_desc_t>>(cgh, [=]() {
+                MKL_LONG status[2] = { DFTI_BAD_DESCRIPTOR };
+
+                for (auto dir : { DIR::fwd, DIR::bwd })
+                    status[dir] = DftiCommitDescriptor(bidir_handle_obj[0][dir]);
+
+                if (status[0] != DFTI_NO_ERROR && status[1] != DFTI_NO_ERROR) {
+                    std::string err = std::string("DftiCommitDescriptor failed with status : ") +
+                                      DftiErrorMessage(status[0]) + std::string(", ") +
+                                      DftiErrorMessage(status[1]);
+                    throw oneapi::mkl::exception("dft/backends/mklcpu", "commit", err);
+                }
```

Comment:
```suggestion
                for (auto dir : { DIR::fwd, DIR::bwd }) {
                    auto status = DftiCommitDescriptor(bidir_handle_obj[0][dir]);

                if (status != DFTI_NO_ERROR) {
                    std::string err = std::string("DftiCommitDescriptor failed with status : ") +
                                      DftiErrorMessage(status);
                    throw oneapi::mkl::exception("dft/backends/mklcpu", "commit", err);
                }
                }
```
I'm not sure we should continue trying after the first one failed

Reply:
no this cannot be done, as in case of real batched transformation, where resetting strides/distances are essential, the 2nd descriptor would be of inconsistent configuration (because of the strides) and would error out.
the first one can be used for compute_forward, then we recommit and reset the strides, which invalidates the first descriptor and commits the second one. 

@hjabird the reason we can't use the same logic to reset the strides is to maintain consistency with the gpu backend (which as it does not have 2 different descriptors does not have that capability)

Reply:
It is a tricky issue.

Best I can suggest is that if 1 of the 2 fail, we save the exception and throw it if the user tries to use that. I'm happy for that to be another PR though.

Reply:
That is a very awkward problem. Please add a comment describing this to your solution and also remember to revert the change from `&&` to `||` if needed.

Reply:
I think there is a similar issue in the cuFFT backend (I added a comment there) and forcing the two descriptors to be valid at the same time might result in failure to commit for valid real FFT problems.

One idea, @raphael-egan, @hpark-intel please let me know if you think of a case this would not work.
Can we try to commit the fwd descriptor with the given input/output stride, if successful commit the bwd descriptor with swapped input/output strides and check that both descriptor are valid?
Similarly if the first commit for forward descriptor fails, try again to commit fwd descriptor with swapped input/output strides and commit backward descriptor with original input/output stride. Fail to commit if both descriptors aren't valid?

Also I agree this might be better addressed in a separate PR.

Reply:
I think it might be a bad idea to swap the input/output strides since they may be designed for one direction but valid in both (NB I'm not 100% sure of this since it's 6:30pm on a Friday). I think a scenario where the data is densely packed in the backward domain but has spaces between the data in the complex domain might do it.

Reply:
@lhuot from my reading, you don't want any changes now so I will resolve this thread.

Reply:
This is indeed a fairly awkward issue that underlines that there is room for major improvement to the specs' and product's APIs: the possibility of having a single set of configuration parameters unambiguously usable for either compute direction would require the `{IN,OUT}PUT_STRIDES` configuration parameters to be deprecated for the SYCL APIs and substituted by (new, yet-to-be-approved-and-implemented) `{F,B}WD_STRIDES` parameters. This is suggested in [MKLD-15060](https://jira.devtools.intel.com/browse/MKLD-15060).

As @FMarno points out, there might be cases where either direction is valid with sufficient padding. Consider a 4x4 real transforms with CCE storage, with input strides {5, 1, 0} and output strides {4,. 1, 0}, the user may be configuring the descriptor for a rfo4:5:4:x4:1:1 (padding of 1 real data point in fwd domain and 1 complex data point in bwd domain) or for a rbo4:5:4:x4:1:1 (padding of 2 complex data points in bwd domain, no padding in fwd domain) and either would be valid. Therefore, swapping strides for the bwd descriptor if the fwd descriptor committed successfully would result in incorrectness if the user actually meant to target rbo4:5:4x4:1:1 in that example.

In my personal opinion, it is best to not swap the strides and allow for one (not both!) of the descriptors to fail to commit and check thereafter that it is never used if it didn't commit successfully prior.